rfcs/45-stageless.md

```diff
@@ -0,0 +1,984 @@
+7. (Optional) You have at least one pair of ambiguously-ordered systems with conflicting data access.
+
+(5), (6), and (7) don't inherently make a graph unsolvable, so they can be configured as ignore, warn, or error.
+By default, they are all configured as warn.
+(7) has additional configuration options.
+See [bevyengine/bevy#4299](https://github.com/bevyengine/bevy/pull/4299) for more details.
+
+### `IntoRunCondition`
+
+```rust
+pub trait IntoRunCondition<Params>: IntoSystem<(), bool, Params> {}
+
+impl<Params, Marker, F> IntoRunCondition<(IsFunctionSystem, Params, Marker)> for F
+where
+    F: SystemParamFunction<(), bool, Params, Marker> + Send + Sync + 'static,
+    Params: SystemParam + 'static,
+    Params::Fetch: ReadOnlySystemParamFetch,
+    Marker: 'static,
+{
+}
+```
+
+Any function that satisfies these constraints can be used as a condition:
+
+- It must return a `bool`.
+- It cannot have system params that grant mutable access to data.
+  - **Reason**: If something has multiple conditions, the order they're evaluated in won't matter.
+
+The following is also strongly-recommended:
+
+- It shouldn't rely on local data (e.g. `Local<T>` or change detection queries).
+  - **Reason**: Again, the order won't matter because they'll all be idempotent (otherwise, you can encounter subtle bugs).
+
+### Evaluating conditions in the proper order
+
+When the executor is presented with a ready system, it does the following:
+
+- Check if access to all the data needed by the system, its conditions, and the conditions of the sets it's under (that haven't been evaluated yet) is available.
+  - If no, leave the system in the ready queue and move onto the next one.
+- If yes, evaluate the conditions of those not-yet-evaluated sets in hierarchical order (from the outermost to the innermost).
+  - If any set's conditions return `false`, mark all the systems (incl. current one) under it as completed and all sets under it as evaluated.
+  - If all of the set's conditions return `true`, mark the set as evaluated.
+- If all those conditions returned `true`, now evaluate the system's conditions.
+  - If any of these return `false`, mark the system as completed.
+- If all the system's conditions returned `true`, run the system.
+  - Mark access to the data required by the system as unavailable.
+  - Create a task for the system.
+  - Spawn and run that task.
+- When the system task completes, refresh the available access.
+
+Each condition is evaluated at most once.
+Since most of these functions are probably simple tests, we don't spawn tasks for them, which avoids locking other systems out of the data they access.
+Likewise, we don't spawn tasks for systems that get skipped. This lightweight approach minimizes task overhead.
+
+### States
+
+The implementation of a basic state machine can be very simple in this new scheme.
+
+```rust
+#[derive(Resource)]
+pub struct CurrentState<S: State>(pub S);

```

Comment:
I think a consolidated `State<GameState>` resource (with current and next fields) is simpler / more discoverable / easier on the eyes.

```rust
fn system(mut game_state: ResMut<State<GameState>>) {
  if *game_state == GameState::InMenu {
     game_state.set_next(GameState::InGame);   
  }
}
```

vs 

```rust
fn system(current_game_state: Res<CurrentState<GameState>>, next_game_state: ResMut<NextState<GameState>>) {
  if *current_game_state == GameState::InMenu {
     *next_game_state = GameState::InGame;   
  }
}
```

Reply:
Agreed, I like this design better.

Reply:
We initially had a consolidated resource, but anecdotally some users said they liked how `iyes_loopless` has them split.

The split just has the advantage that one system can queue a transition without blocking another system that just wants to read the current state.

Reply:
Oh, one more important property with split resources is that `CurrentState` wouldn't be marked as changed when you queue a transition. (bevyengine/bevy#2343)

Reply:
The change detection argument is very compelling to me.

Reply:
Hmmm ok yeah theres serious value in being able to detect state changes. Can we change it to `State<GameState>` instead of `CurrentState<GameState>` for some additional terseness? I think "current" is implied.

Alternatively, we could just use interior mutability for `State::set_next()`. Given that state transitions will happen approximately once per frame, the overhead (even if we use a mutex) won't matter at all.

Reply:
I like the rename. I don't love the interior mutability though because it breaks change detection for NextState. 

Less essential, but likely still useful.

Reply:
> Can we change it to `State<GameState>` instead of `CurrentState<GameState>` for some additional terseness?

Sure, just need to keep the trait name as `StateData` or `StateType`.

Reply:
Hmm sounds like you were one step ahead of me here. `#[derive(State)]` does make sense to me, as `struct GameState` "is" the State. I was thinking within the context of the current approach, which doesn't use a State trait directly. But `State<GameState>` _is_ the ideal reading for the system param imo.

We've been able to sidestep this issue in the past because other system params use different names than their "contained" types: `Res` vs `Resource`, `&X` vs `Component`, `Query` vs `WorldQuery`.

I guess on main we currently use `StateData` as the trait and `State` as the param. Thats not the worst thing in the world. Idk I don't have strong opinions here anymore :)

Reply:
Going with the existing design, and marking this as resolved.

We can swap this later if needed.